### PR TITLE
Avoid crashing kubelet when trying to use the RuntimeClassManager wit…

### PR DESCRIPTION
**Pull request by aholler**
…hout an api-server

Crashing kubelet makes it unnecessarily hard to find the real problem.

E.g. in my case it was the case that the default system unit file for
kubelet on Gentoo ignores the config provided by kubeadm. Therefore calling
kubeadm join ended up with starting kubelet without a config which then
resulted in a crash when the RuntimeClassManager tried to access the
non-existent kubeconfig.

Avoid that.

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: pkg/kubelet/runtimeclass/runtimeclass_manager.go [hunk 1]
```diff
@@ -35,6 +35,10 @@ type Manager struct {
 func NewManager(client clientset.Interface) *Manager {
 	const resyncPeriod = 0
 
+	if client == nil {
```
Comment (yastij) [inline pkg/kubelet/runtimeclass/runtimeclass_manager.go]: While I agree that we should fix this, I'm not in favor of failing silently. I think we should `return nil, err` and handle the error when constructing `KubeletConfiguration`

**Review by yastij — CHANGES_REQUESTED** [review]

**Comment by aholler** [comment]
Sorry,

but I don't know enough about how kubelet works internally to implement 
a better error handling for that. I've seen that other plugins are 
handling the same case directly in kubelet.go (also without doing more 
than my commit), but I've choosen to put that if-statement in 
runtimeclass_manager.go in order keep it where it belongs to.

If necessary, I can add a klog-call with some message, but for 
everything else, someone else has to provide a better fix.

Best regards,

Alexander

Am 05.05.2019 um 13:11 schrieb Yassine TIJANI:

**Comment by aholler** [comment]
I've signed the CLA.

Regards,

Alexander

Am 05.05.2019 um 09:38 schrieb Kubernetes Prow Robot:

**Comment by aholler** [comment]
Just as a side note, The feature RuntimeClass got only enabled by default with 1.14. Therefore many people likely haven't ended up with that problem (more will come). But besides other unclear descriptions of a similar (or maybe the same) problem, there is the issue  #76424 which would have been fixed by that commit here too.

File: pkg/kubelet/runtimeclass/runtimeclass_manager.go (hunk 1 above)
Comment (mattjmcnaughton) [reply pkg/kubelet/runtimeclass/runtimeclass_manager.go]: +1 to returning `nil, err`. 

Re what to do with that error...

One option would be to log the error in the `NewMainKubelet` function in `pkg/kubelet/kubelet.go` which calls `runtimeclass.NewManager`, and then continue processing. Alternatively, we could have `NewMainKubelet` return an error if `runtimeclass.NewManager` returns an error.

Curious for the communities opinion on whether `RuntimeClassManager` failing to start should cause the entire kubelet to fail. It seems like `NewMainKubelet` has a mix of failing/continuing with execution if individual managers fail.

**Review by mattjmcnaughton — COMMENTED** [review]
Thanks for your pr :)

As a total side note... respect to you for running k8s on Gentoo :1st_place_medal:

**Comment by mattjmcnaughton** [comment]
Oh also, if your pr does not require a release note (and I agree yours doesn't), please specify the release note as `NONE`.

**Comment by aholler** [comment]
Hello,

I'm sorry, but I don't see why returning an error should help.
That would have to been either ignored or translated into some meaningful message for the user, neither of which I think is necessary.

Besides that I'm sorry but I have to draw a line here.
I had the obvious silly idea someone would be happy about a simple, well documented "one-line"-patch and I wouldn't have to spend endless hours to discuss something, in my eyes, very simple.

And now, I already had to sign some CLA, received around a dozen mails and ...

Therefore, please take it and maybe put a patch on top with whatever changes you like, or just fix it yourself however you want or even ignore it. I don't care much and I won't become angry, regardless which solution will be choosen.

I've spend the time to analyze and fix that little bug, but I don't want to spend much more time to discuss that little patch.

Best regards,

Alexander

**Comment by mattjmcnaughton** [comment]
> Hello,
> 
> I'm sorry, but I don't see why returning an error should help.
> That would have to been either ignored or translated into some meaningful message for the user, neither of which I think is necessary.
> 
> Besides that I'm sorry but I have to draw a line here.
> I had the obvious silly idea someone would be happy about a simple, well documented "one-line"-patch and I wouldn't have to spend endless hours to discuss something, in my eyes, very simple.
> 
> And now, I already had to sign some CLA, received around a dozen mails and ...
> 
> Therefore, please take it and maybe put a patch on top with whatever changes you like, or just fix it yourself however you want or even ignore it. I don't care much and I won't become angry, regardless which solution will be choosen.
> 
> I've spend the time to analyze and fix that little bug, but I don't want to spend much more time to discuss that little patch.
> 
> Best regards,
> 
> Alexander

Thanks for your work finding the bug @aholler :) We really appreciate it.

Thanks for being upfront about your decision - feel free to close your PR and I can create another one where we can sort out the remaining open questions.

I hope we'll get this merged and fixed soon and thanks again for your help :)

**Comment by mattjmcnaughton** [comment]
Created https://github.com/kubernetes/kubernetes/pull/77778 to carry on this work :) Thanks again Alexander!

**Comment by tallclair** [comment]
Fixed by https://github.com/kubernetes/kubernetes/pull/77888

**Comment by aholler** [comment]
Thanks for fixing it in a form you like.

**Pull request closed without merging**